language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -55,48 +124,112 @@ type VarEnv = UniqueMap TermUnique Closure
 data CekEnv = CekEnv
     { _cekEnvMeans  :: DynamicBuiltinNameMeanings
     , _cekEnvVarEnv :: VarEnv
+    , _cekEnvProtocolParameters :: ()
     }
 
+data ExBudget = ExBudget { _exBudgetCPU :: ExCPU, _exBudgetMemory :: ExMemory }
+
+-- TODO memory should probably be fixed size - but then how do we handle overflow?
+newtype ExMemory = ExMemory Integer -- Probably machine word size
+  deriving (Eq, Ord)
+  deriving newtype Num
+  deriving (Semigroup, Monoid) via (Sum Integer)
+newtype ExCPU = ExCPU Integer
+  deriving (Eq, Ord)
+  deriving newtype Num
+
+makeLenses ''ExBudget
+
 -- | The monad the CEK machine runs in.
-type CekM = ReaderT CekEnv (Either CekMachineException)
+type CekMBase w s = ExceptT CekMachineException (RWS CekEnv w s)
+type SpendExUnits w s = (SpendExUnitsState s, SpendExUnitsWriter w)
+type CekM w s a = SpendExUnits w s => CekMBase w s a
+
+data ExTally = ExTally
+    { _exTallyCPU :: MonoidalHashMap (Plain Term) [ExCPU]
+    , _exTallyMemory :: MonoidalHashMap (Plain Term) [ExMemory]
+    }
+    deriving stock (Eq, Generic)
+    deriving (Semigroup, Monoid) via (GenericSemigroupMonoid ExTally)
+
+spendBoth :: WithMemory Term -> ExCPU -> ExMemory -> CekM w s ()
+spendBoth term cpu mem = spendCPU term cpu >> spendMemory term mem
+
+spendMemory :: WithMemory Term -> ExMemory -> CekM w s ()
+spendMemory term mem = do
+    spendSMemory term mem
+    spendWMemory term mem
+
+spendCPU :: WithMemory Term -> ExCPU -> CekM w s ()
+spendCPU term mem = do
+    spendSCPU term mem
+    spendWCPU term mem
+
+class Monoid w => SpendExUnitsWriter w where
+    spendWMemory :: forall s. WithMemory Term -> ExMemory -> CekMBase w s ()
```

Comment:
I think it would be perhaps more natural to have this be parameterized over the monad too. Then in your instance for `ExBudget` you can just require some appropriate `MonadError`/`MonadState` superclass constraints.

Although that's pushing in the direction of making this more MTL-y, which maybe we don't want.

Reply:
MTL has been a bit of a headache for me, hence the use of a specific stack and more flexible variables. That way the control flow is set in stone.